applications/StructuralMechanicsApplication/custom_strategies/custom_strategies/harmonic_analysis_strategy.hpp

```diff
@@ -0,0 +1,743 @@
+
+        if (BaseType::GetEchoLevel() > 2 && rank == 0)
+            std::cout << "Exiting InitializeSolutionStep() of HarmonicAnalysisStrategy" << std::endl;
+
+        KRATOS_CATCH("")
+    }
+
+    /// Check whether initial input is valid.
+    virtual int Check()
+    {
+        KRATOS_TRY
+
+        auto& rModelPart = BaseType::GetModelPart();
+        const auto rank = rModelPart.GetCommunicator().MyPID();
+
+        if (BaseType::GetEchoLevel() > 2 && rank == 0)
+            std::cout << "Entering Check() of HarmonicAnalysisStrategy" << std::endl;
+
+        // check the model part
+        BaseType::Check();
+
+        // check the scheme
+        this->pGetScheme()->Check(rModelPart);
+
+        // check the builder and solver
+        this->pGetBuilderAndSolver()->Check(rModelPart);
+
+        if (BaseType::GetEchoLevel() > 2 && rank == 0)
+            std::cout << "Exiting Check() of HarmonicAnalysisStrategy" << std::endl;
+
+        return 0;
+
+        KRATOS_CATCH("")
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+    ///@}
+    ///@name Friends
+    ///@{
+
+    ///@}
+
+protected:
+    ///@name Protected static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected member Variables
+    ///@{
+
+    ///@}
+    ///@name Protected Operators
+    ///@{
+
+    ///@}
+    ///@name Protected Operations
+    ///@{
+
+    ///@}
+    ///@name Protected  Access
+    ///@{
+
+    ///@}
+    ///@name Protected Inquiry
+    ///@{
+
+    ///@}
+    ///@name Protected LifeCycle
+    ///@{
+
+    ///@}
+
+private:
+    ///@name Static Member Variables
+    ///@{
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    SchemePointerType mpScheme;
+
+    BuilderAndSolverPointerType mpBuilderAndSolver;
+
+    bool mInitializeWasPerformed;
+
+    SparseVectorPointerType mpForceVector;
+
+    DenseMatrixPointerType mpModalMatrix;
+
+    double mRayleighAlpha;
+
+    double mRayleighBeta;
+
+    double mSystemDamping;
+
+    bool mUseMaterialDamping;
+
+    vector< double > mMaterialDampingRatios;
+
+    ///@}
+    ///@name Private Operators
+    ///@{
+
+    ///@}
+    ///@name Private Operations
+    ///@{
+
+    /// Assign the modal displacement to the dofs and the phase angle to the reaction
+    void AssignVariables(ComplexVectorType& rModalDisplacement, int step=0)
+    {
+        auto& rModelPart = BaseType::GetModelPart();
+        for( auto itNode = rModelPart.NodesBegin(); itNode != rModelPart.NodesEnd(); itNode++ )
+        {
+            ModelPart::NodeType::DofsContainerType& rNodeDofs = itNode->GetDofs();
+            
+            for( auto itDof = std::begin(rNodeDofs); itDof != std::end(rNodeDofs); itDof++ )
+            {
+                if( !itDof->IsFixed() )
+                {
+                    itDof->GetSolutionStepValue(step) = std::abs(rModalDisplacement(itDof->EquationId()));
```

Comment:
i don't really understand the meaning of applying the same value to the reaction and to the value.

Reply:
I apply the absolute displacement on the solution step value and the phase angle (std::arg) on the reaction.